HarbourMasters/Shipwright
# CrowdControl V2: Effect overhaul

Whenever we release the SoH version that includes this, we have to communicate with the CC team to upload the .cs file on their end. They'll also have to add all the effect descriptions and prices which can be found in the .ccpak file, and update the images for the effects accordingly.

Thanks to Caladius for being a sparring partner & helping me with some tedious stuff.
Thanks to RR for lending me their Random Bonk While Rolling code.
Thanks to dcvz for helping me get GameInteractor out the door that made all of this much easier to implement.

To do:
- [x] Testing + fixes
- [x] Fix Linux build
- [x] Finish adding prices and descriptions to effects

New effects:
- Spawn Lit Bomb
- Spawn Explosion
- Spawn Gibdo
- Spawn Iron Knuckle
- Spawn Flock of Keese (5x), replaces Spawn Keese
- Take Half Damage (replaces Defense Modifier)
- Take Double Damage (replaces Attack Modifier)
- Invincibility
- Slippery Floor
- Disable Ledge Grabs
- Random Wind
- Random Bonks When Rolling
- Give/Take Shields (Deku & Hylian)
- Refill/Take Consumables
- Squished Link
- Random Bomb Fuse Timer
- Set Time to Dawn/Dusk
- Debug Mode
- Randomize Cosmetics
- Press Random Buttons
- Clear C-Buttons or D-Pad
- Teleport player to Link's House/Warp Song Destinations
- Bid wars for Tunic/Navi/Link's Hair color

Removed Effects:
- Spawn Tektite (felt pointless)

Other features:
- Completely revamped effect layout (everything is in folders now)
- Added descriptions to effects
- New standard prices
- Enemies spawn in a circle around player in a random direction
- Enemies get put onto the ground a-la Enemy Rando and have an extra OoB check
- Put strings into map and switch by enums instead, making the code much easier to read.

Bugfixes:
- Fix Link's size effects bugging out doors, ladders and crawlspaces (temporarily resets size during said actions)
- Gravity levels no longer make Link sink in water
- Disable enemy spawning in Forest Temple Poe rooms to prevent softlocks https://github.com/HarbourMasters/Shipwright/issues/2333
- Fix Iron Knuckles not properly respawning in Spirit Temple
- Limit Dark Links to 1 maximum at a time to prevent shared static variables making Dark Links affect each other.
- As per the GameInteractor PR, many effects have improved checks for when an effect should be possible or not.


### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173080.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173081.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173082.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173083.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173085.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/581173086.zip)
